src/cryptonote_protocol/request_manager.cpp

```diff
@@ -0,0 +1,255 @@
+  std::chrono::steady_clock::time_point now = std::chrono::steady_clock::now();
+  MINFO("Adding peer: " << epee::string_tools::pod_to_hex(id)
+        << " to transaction: "
+        << epee::string_tools::pod_to_hex(tx_hash));
+  std::unique_lock<std::recursive_mutex> lock(m_mutex);
+  auto& by_peer_and_tx = get_requests_by_peer_and_tx(m_requested_txs);
+  auto it = by_peer_and_tx.find(boost::make_tuple(id, tx_hash));
+  if (it != by_peer_and_tx.end()) {
+    // already have this peer for this tx
+    MDEBUG("Peer " << epee::string_tools::pod_to_hex(id)
+           << " already in request queue for tx "
+           << epee::string_tools::pod_to_hex(tx_hash));
+    return false;
+
+  }
+  m_requested_txs.insert(tx_request(id, tx_hash, now));
+  return true;
+}
+
+void request_manager::add_request(const crypto::hash &tx_hash, const boost::uuids::uuid &peer_id) {
+  MINFO("Requesting from peer: " << epee::string_tools::pod_to_hex(peer_id)
+        << " the transaction: "
+        << epee::string_tools::pod_to_hex(tx_hash));
+  std::unique_lock<std::recursive_mutex> lock(m_mutex);
+  // Search directly in m_requested_txs instead of using the index
+  for(auto& req : m_requested_txs) {
+      if(req.peer_id == peer_id && req.tx_hash == tx_hash) {
+          req.fly();
+          return;
+      }
+  }
+  MINFO("Requested tx " << tx_hash << " that does not exist in request_manager queue.");
+}
+
+bool request_manager::get_current_request_peer_id(const crypto::hash &tx_hash, boost::uuids::uuid &out) const {
+  std::unique_lock<std::recursive_mutex> lock(m_mutex);
+  auto& by_tx_hash = get_requests_by_tx_hash(m_requested_txs);
+  auto it = by_tx_hash.find(tx_hash);
+  if (it == by_tx_hash.end()) {
+    MDEBUG("No requests for tx " << epee::string_tools::pod_to_hex(tx_hash));
+    return false;
+  }
+  for (const auto& req : boost::make_iterator_range(by_tx_hash.equal_range(tx_hash))) {
+    if (req.is_in_flight()) {
+      out = req.peer_id;
+      return true;
+    }
+  }
+  return false;
+}
+
+boost::uuids::uuid request_manager::request_from_next_peer(const crypto::hash &tx_hash, std::chrono::steady_clock::time_point now) {
+  std::unique_lock<std::recursive_mutex> lock(m_mutex);
+  auto& by_tx_hash = get_requests_by_tx_hash(m_requested_txs);
+  auto it = by_tx_hash.find(tx_hash);
+  if (it == by_tx_hash.end()) {
+    MDEBUG("No requests for tx " << epee::string_tools::pod_to_hex(tx_hash));
+    return boost::uuids::nil_uuid();
+  }
+
+  // remove in flight request that have timed out
+  auto range = by_tx_hash.equal_range(tx_hash);
+  it = range.first;
+  while (it != range.second) {
+    if (it->is_in_flight()) {
+      MINFO("Erasing " << epee::string_tools::pod_to_hex(tx_hash)
+            << " from peer "
+            << epee::string_tools::pod_to_hex(it->peer_id));
+      it->mark_for_removal();
+    } else {
+      ++it;
+    }
+  }
+
+  // Recompute range after potential erasures.
+  by_tx_hash = get_requests_by_tx_hash(m_requested_txs);
```

Comment:
I don't understand what's going on at this line. `by_tx_hash` is a reference to something, are you reassigning it? It's an unconventional use of C++ references, so I'm confused. Maybe it's better to give it a new name `by_tx_hash2` or similar to not mix with the above code?